### PR TITLE
PayPal Standard: Refactor order address update functionality

### PR DESCRIPTION
**Pull request by Mayisha**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:
- We were updating the WC order's addresses in the callback of `woocommerce_before_thankyou` hook. In this callback, we retrieve the order details from PayPal and store the addresses from the response to WC order. If the callback is triggered multiple times (by the same hook used by other extensions, by thank you page reload or for any network issue on the thank you page) this callback method to update the address will be called multiple times.
    - Added a new meta in the order `_paypal_address_update_status` to detect whether the address update has been attempted or not, and the status.
- As the address update depends on the get order details response from PayPal, there could be cases when this API request fails and as a result the address is not updated in WC order. For these edge case scenarios, we are updating the address when we receive the `CHECKOUT.ORDER.APPROVED` event as well because this event data also has the addresses.
    - We update the address from here only if the address update did not take place in the previous hook.
    - On the thank you page, we show shoppers' order details and addresses. If we only update the address in the webhook handler, shoppers' won't see the correct address on thank you page. So this is just a fallback for the edge case.
- Moved the address update function to the helper class to reuse it in two places.

Closes PAYPAL-137

### How to test the changes in this Pull Request:
- Enable PayPal Standard `wp option patch update woocommerce_paypal_settings _should_load 'yes'`
- In [get_paypal_order_details method](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/gateways/paypal/includes/class-wc-gateway-paypal-request.php#L231), add a test log to detect the method call.
- As a shopper, add a product to your cart and go to the checkout page.
- Select PayPal Standard as the payment method and checkout.
- On the PayPal modal, change the shipping address and complete payment.
- Check your logs and confirm that `get_paypal_order_details` was called once.
- Reload the thank you page and verify that the `get_paypal_order_details` was not called again.
- As an admin, check the order details page and verify that the selected shipping address from PayPal modal is saved as the shipping address.
- In [get_paypal_order_details method](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/gateways/paypal/includes/class-wc-gateway-paypal-request.php#L231), keep the previous log and change the `$paypal_order_id` (example: $paypal_order_id='invalid'). This will ensure the address update fails in the `woocommerce_before_thankyou` callback and will be attempted when webhook is received.
- As a shopper, add a product to your cart and go to the checkout page.
- Select PayPal Standard as the payment method and checkout.
- On the PayPal modal, change the shipping address and complete payment.
- Check your logs and confirm that `get_paypal_order_details` was called once.
- Reload the thank you page and verify that the `get_paypal_order_details` was not called again.
- As an admin, check the order details page and verify that the selected shipping address from the PayPal modal is saved as the shipping address. (this may take a few seconds depending on webhook event arrival)

### Milestone

- [x] Automatically assign milestone for the **[next WooCommerce version](../blob/trunk/plugins/woocommerce/woocommerce.php#L6)**

> **Note:** Check the box above to have the milestone automatically assigned when merged.
> Alternatively (e.g. for point releases), manually assign the appropriate milestone.

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 
Refactored the address update functionality of PayPal Standard with improved status tracking to prevent duplicate requests. 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/includes/gateways/paypal/includes/class-wc-gateway-paypal-helper.php [hunk 1]
```diff
@@ -152,4 +152,64 @@ public static function mask_email( $email ) {
...
+	 * @return void
+	 */
+	public static function update_addresses_in_order( ?WC_Order $order, array $paypal_order_details ): void {
+		if ( empty( $order ) || empty( $paypal_order_details ) ) {
+			return;
+		}
+
+		// Bail early if the addresses update already have been successful.
+		$addresses_update_status = $order->get_meta( '_paypal_address_update_status', true );
+		if ( 'success' === $addresses_update_status ) {
```
Comment (wjrosa) [inline plugins/woocommerce/includes/gateways/paypal/includes/class-wc-gateway-paypal-helper.php]: If we have just 2 statuses for this, I would consider converting it to a boolean flag

File: plugins/woocommerce/includes/gateways/paypal/includes/class-wc-gateway-paypal-helper.php [hunk 2]
```diff
@@ -152,4 +152,64 @@ public static function mask_email( $email ) {
...
+		}
+
+		// Update the shipping information.
+		$full_name = $paypal_order_details['purchase_units'][0]['shipping']['name']['full_name'] ?? '';
+		if ( ! empty( $full_name ) ) {
+			$approximate_first_name = explode( ' ', $full_name )[0] ?? '';
+			$approximate_last_name  = explode( ' ', $full_name )[1] ?? '';
+			$order->set_shipping_first_name( $approximate_first_name );
+			$order->set_shipping_last_name( $approximate_last_name );
+		}
```
[the remark this answers is not in the document]
Comment (wjrosa) [reply plugins/woocommerce/includes/gateways/paypal/includes/class-wc-gateway-paypal-helper.php:181]: Another idea to solve this issue would be to use `end( $name_parts )`.

**Review by wjrosa — APPROVED** [review]
Looks good to me! Left some suggestions

**Review by diegocurbelo — APPROVED** [review]
The code changes look good to me, and it tests well!

File: plugins/woocommerce/includes/gateways/paypal/includes/class-wc-gateway-paypal-helper.php [hunk 3]
```diff
@@ -152,4 +152,64 @@ public static function mask_email( $email ) {
...
 	}
+
+	/**
+	 * Update the addresses in the order.
+	 *
+	 * @param WC_Order $order The order object.
+	 * @param array    $paypal_order_details The PayPal order details.
+	 * @return void
+	 */
+	public static function update_addresses_in_order( WC_Order $order, array $paypal_order_details ): void {
```
[the remark this answers is not in the document]
Comment (Mayisha) [reply plugins/woocommerce/includes/gateways/paypal/includes/class-wc-gateway-paypal-helper.php]: Adding the helper method is simpler and more helpful here.

File: plugins/woocommerce/includes/gateways/paypal/includes/class-wc-gateway-paypal-helper.php (hunk 1 above)
Comment (Mayisha) [reply plugins/woocommerce/includes/gateways/paypal/includes/class-wc-gateway-paypal-helper.php]: Done.

File: plugins/woocommerce/includes/gateways/paypal/includes/class-wc-gateway-paypal-webhook-handler.php [hunk 4]
```diff
@@ -92,6 +92,9 @@ private function process_checkout_order_approved( $event ) {
 			);
 			$order->save();
 
+			// Update the addresses in the order with the addresses from the PayPal order details.
+			WC_Gateway_Paypal_Helper::update_addresses_in_order( $order, $event['resource'] );
```
[the remark this answers is not in the document]
Comment (Mayisha) [reply plugins/woocommerce/includes/gateways/paypal/includes/class-wc-gateway-paypal-webhook-handler.php:96]: The webhook event brings back the correct address in case shopper had changed the address on PayPal modal. This update is important.

File: plugins/woocommerce/includes/gateways/paypal/includes/class-wc-gateway-paypal-helper.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (Mayisha) [reply plugins/woocommerce/includes/gateways/paypal/includes/class-wc-gateway-paypal-helper.php:181]: Addressed.

File: plugins/woocommerce/includes/gateways/paypal/includes/class-wc-gateway-paypal-helper.php [hunk 5]
```diff
@@ -152,4 +152,64 @@ public static function mask_email( $email ) {
...
+	 * @param WC_Order|null $order The order object.
+	 * @param array         $paypal_order_details The PayPal order details.
+	 * @return void
+	 */
+	public static function update_addresses_in_order( ?WC_Order $order, array $paypal_order_details ): void {
+		if ( empty( $order ) || empty( $paypal_order_details ) ) {
+			return;
+		}
+
+		// Bail early if '_paypal_addresses_updated' is 'yes', meaning the addresses update already have been successful.
```
[the remark this answers is not in the document]
Comment (Mayisha) [reply plugins/woocommerce/includes/gateways/paypal/includes/class-wc-gateway-paypal-helper.php:168]: The changes with previous meta key was not merged.
